### PR TITLE
fix(base-cluster/backup)!: only enable velero if a backupstoragelocation is set

### LINKED ISSUE #752: [base-cluster] only install velero when a default storage location is defined
Otherwise velero won't start. 

I'm so glad that we use velero.

### PR DESCRIPTION
**Pull request by tasches**
closes #752

### CODE REVIEW

**Comment by marvinWolff** [comment]
This looks fine to me.

But i'm asking myself how the rollout will go.

Most clusters currently aren't using velero, but some do. In those clusters the backupStorageLocation ist not set via the chart. At least not that i'm aware of.  
Here we should either move the declaration into the chart or add some sort of configuration option ala `existingStorageLocation: <name>`.

But im not sure how practical that would be. What do you think @cwrau @tasches ?

File: charts/base-cluster/templates/backup/validation.yaml [hunk 1]
```diff
@@ -3,3 +3,9 @@
     {{- fail (printf "The `defaultLocation` '%s' must exist in `backupStorageLocations`" .Values.backup.defaultLocation) -}}
   {{- end -}}
 {{- end -}}
+{{- if .Values.backup.enabled -}}
+  {{- $length := len .Values.backup.backupStorageLocations -}}
+  {{- if eq $length 0 -}}
+    {{- fail (print "At least one .Values.backup.backupstoragelocations needs to be configured")  -}}
+  {{- end -}}
+{{- end -}}
```
Comment (cwrau) [inline charts/base-cluster/templates/backup/validation.yaml:11]: You should replace this whole block with the following inside the schema;

```diff
diff --git a/charts/base-cluster/values.schema.json b/charts/base-cluster/values.schema.json
index 18c64f5..62b367b 100644
--- a/charts/base-cluster/values.schema.json
+++ b/charts/base-cluster/values.schema.json
@@ -1278,7 +1278,8 @@
               "bucket"
             ],
             "additionalProperties": false
-          }
+          },
+          "minProperties": 1
         },
         "defaultLocation": {
           "type": "string"
@@ -1293,6 +1294,9 @@
           "additionalProperties": false
         }
       },
+      "required": [
+        "defaultLocation"
+      ],
       "additionalProperties": false
     },
     "common": {
```

which should accomplish the same

File: charts/base-cluster/values.yaml [hunk 2]
```diff
@@ -443,7 +443,7 @@ rbac:
   accounts: {}
 
 backup:
-  enabled: true
+  enabled: false
```
Comment (cwrau) [inline charts/base-cluster/values.yaml:446]: This is a breaking change, which is fine, you just need to change your commit message to include an `!`; `fix(base-cluster)!:`

And you should include the subarea in you message; `fix(base-cluster/backup)!:`

**Review by cwrau — DISMISSED** [review]

File: charts/base-cluster/templates/backup/validation.yaml (hunk 1 above)
Comment (tasches) [reply charts/base-cluster/templates/backup/validation.yaml]: no longer necessary

**Comment by cwrau** [comment]
You're missing some locations where `.backup.enabled` is used;

```
charts/base-cluster/values.yaml
charts/base-cluster/templates/backup/velero.yaml
charts/base-cluster/templates/backup/validation.yaml
charts/base-cluster/templates/backup/migrations/velero-4-to-5.yaml
charts/base-cluster/templates/monitoring/kube-prometheus-stack/_grafana-config.yaml
```

You can just search for `.backup.enabled`

**Comment by cwrau** [comment]
You should require `defaultLocation` if `backupStorageLocations` is defined;

```diff
diff --git a/charts/base-cluster/values.schema.json b/charts/base-cluster/values.schema.json
index d8f2f69..948cc3b 100644
--- a/charts/base-cluster/values.schema.json
+++ b/charts/base-cluster/values.schema.json
@@ -1290,6 +1290,25 @@
           "additionalProperties": false
         }
       },
+      "oneOf": [
+        {
+          "properties": {
+            "backupStorageLocations": {
+              "minProperties": 1
+            }
+          },
+          "required": [
+            "defaultLocation"
+          ]
+        },
+        {
+          "properties": {
+            "backupStorageLocations": {
+              "maxProperties": 0
+            }
+          }
+        }
+      ],
       "additionalProperties": false
     },
     "common": {
```

In this case I'd also be fine with the `if`/`then`;

```diff
diff --git a/charts/base-cluster/values.schema.json b/charts/base-cluster/values.schema.json
index d8f2f69..4ac7ee7 100644
--- a/charts/base-cluster/values.schema.json
+++ b/charts/base-cluster/values.schema.json
@@ -1290,6 +1290,18 @@
           "additionalProperties": false
         }
       },
+      "if": {
+        "properties": {
+          "backupStorageLocations": {
+            "minProperties": 1
+          }
+        }
+      },
+      "then": {
+        "required": [
+          "defaultLocation"
+        ]
+      },
       "additionalProperties": false
     },
     "common": {
```

File: charts/base-cluster/values.yaml [hunk 3]
```diff
@@ -56,7 +56,7 @@ global:
         app.kubernetes.io/component: storage
         app.kubernetes.io/part-of: nfs-server-provisioner
     backup:
-      condition: "{{ .Values.backup.enabled }}"
+      condition: "{{ not (empty .Values.backup.backupStorageLocations) }}"
```
Comment (marvinWolff) [inline charts/base-cluster/values.yaml:59]: Is this `not empty` really needed?

I'm unsure. We haven't used it anywhere else, but i think it also won't hurt?

Comment (marvinWolff) [reply charts/base-cluster/values.yaml:59]: Alright, was cleared. It's ok

**Review by marvinWolff — APPROVED** [review]

**Pull request merged**